loopy-commands.el

```diff
@@ -701,18 +701,51 @@ whose value is to be accumulated."
                   `(loopy--implicit-accumulation-final-update
                     . (setq ,value-holder (nreverse ,value-holder)))
                   `(loopy--implicit-return . ,value-holder)))))
-          (prepend
+          ((prepend prepending)
            `((loopy--main-body
               . (setq ,value-holder (nconc ,value-expression
                                            ,value-holder)))
              (loopy--implicit-return . ,value-holder)))
-          ((push-into push)
-           `((loopy--main-body . (push ,value-expression ,value-holder))
-             (loopy--implicit-return . ,value-holder)))
-          (sum
-           `((loopy--main-body
-              . (setq ,value-holder (+ ,value-holder ,value-expression)))
-             (loopy--implicit-return . ,value-holder)))))))
+          ;; Feed back into `loopy--parse-accumulation-commands' to avoid
+          ;; duplicating code.  We remove any setting of accumulation vars,
+          ;; which we already do above.
+          (t
+           (cl-remove-if (lambda (x) (eq (car x) 'loopy--accumulation-vars))
+                         (loopy--parse-accumulation-commands
+                          (list name value-holder value-expression))))))))
+
+;;;; Boolean Commands
+(cl-defun loopy--parse-always-command ((_ &rest conditions))
+  "Parse a command of the form `(always [CONDITIONS])'.
+If any condition is nil, `loopy' should immediately return nil.
+Otherwise, `loopy' should return t."
+  `((loopy--after-do . (cl-return t))
+    ;; Check all conditions at the end of the loop body, forcing an exit if any
+    ;; evaluate to nil.  Since the default return value of the macro is nil, we
+    ;; don’t need to do anything else.
+    ;;
+    ;; NOTE: We must not add anything to `loopy--final-return', since that
+    ;;       would override the value of any early returns.
+    ,@(mapcar (lambda (condition) `(loopy--post-conditions . ,condition))
+	      conditions)))
+
+(cl-defun loopy--parse-never-command ((_ &rest conditions))
+  "Parse a command of the form `(never [CONDITIONS])'.
+If any condition is t, `loopy' should immediately return nil.
+Otherwise, `loopy' should return t."
+  `((loopy--after-do . (cl-return t))
+    ,@(mapcar (lambda (condition) `(loopy--post-conditions . (not ,condition)))
+	      conditions)))
+
+(cl-defun loopy--parse-thereis-command ((_ &rest conditions))
+  "Parse a command of the form `(thereis [CONDITIONS]).'
+If any condition is non-nil, its value is immediately returned and the loop is exited.
+Otherwise the loop continues and nil is returned."
+  (let ((thereis-var (gensym "thereis-var-")))
+    `((loopy--after-do . (cl-return nil))
+      ,@(mapcar (lambda (condition)
+		  `(loopy--post-conditions . (if-let (,thereis-var ,condition) (cl-return ,thereis-var) t)))
+		conditions))))
```

Comment:
- Instead of using `post-conditions`, I think that it would be better to use something like `(cl-return-from ,loopy--loop-name ,VALUE)`.  That way, the loop is exited as soon as the command's expansion is evaluated, instead of exiting at the end of the loop. This would avoid problems if one is mutating a variable in the loop after running these commands.
- Instead of inserting code into the `else-do` section, I think that it might be better to use an implicit return value like we do with accumulation commands. While the example of an `always` command in the Org documentation inserts code into `else-do`, I suspect that setting a return value implicitly would lead to less surprises. The feature of implicit return values was added after that example was written.

Reply:
A question: I used the argument `(&rest conditions)` to follow the `always` command example, but I don't understand why an accumulation command would take multiple conditions when the user could just pass in `(and condition1 conditon2 ...)` as a condition. Should I just make it one condition: `(condition)`?